plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/rule/InMemoryRuleProcessingService.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.plugin.wlm.rule;
+
+import org.opensearch.plugin.wlm.rule.attribute_extractor.AttributeExtractor;
+import org.opensearch.plugin.wlm.rule.storage.AttributeValueStore;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * This class is responsible for managing in-memory view of Rules and Find matching Rule for the request
+ */
+public class InMemoryRuleProcessingService {
+
+    /**
+     * Default constructor
+     */
+    public InMemoryRuleProcessingService() {}
+
+    /**
+     * Adds the rule to in-memory view
+     * @param rule to be added
+     */
+    public void add(final Rule rule) {
+        new AddRuleOperation(rule).perform();
```

Comment:
I am assuming that the add method has addMerge semantics, so it will take care of updates as well?

Reply:
Yes, So the update will translate into `delete + add`

Reply:
I assume its the user/client responsibility to fetch the latest Rule API via GET, update it with desired params, and then call UPDATE API. And here we will do `delete + add`? Is this is the behaviour?

Shouldn't we still provide our own implementation of UPDATE, as what if there concurrent updates to rule API which ends in inconsistent data? 
We can solve this by either doing delete/add in synchronized block or introducing versioning within the Rule object, where a user update call only succeeds when there no addition updates in between.

Reply:
Good point!, I think versioning would be an overkill since it burdens rule object with an additional field just for the sake of concurrency handling. Additionally with versioning we might end up relying on user to supply the version value to detect the concurrent update correctly.

Reply:
Versioning is one nice/easy way to solve this but not the only one. We also use versioning in OpenSearch for doc updates. But synchronized block approach also works, where you basically process concurrent updates in a serial manner.